File: Backend/app/schemas/product.py
"""Product + ProductCategory schemas — design doc §3 item 2, §5.3."""
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ProductType
from app.schemas.common import Money


class ProductCategoryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)


class ProductCategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    is_active: bool


class ProductCreate(BaseModel):
    name: str = Field(min_length=1, max_length=128)
    product_type: ProductType
    category_id: int | None = None
    sales_price: Money = Decimal("0")
    purchase_cost: Money = Decimal("0")
    default_tax_rate: Money = Decimal("0")
    income_account_id: int | None = None
    expense_account_id: int | None = None


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    product_type: ProductType | None = None
    category_id: int | None = None
    sales_price: Money | None = None
    purchase_cost: Money | None = None
    default_tax_rate: Money | None = None
    income_account_id: int | None = None
    expense_account_id: int | None = None


class ProductOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    product_type: ProductType
    category_id: int | None
    category: ProductCategoryOut | None
    sales_price: Money
    purchase_cost: Money
    default_tax_rate: Money
    income_account_id: int | None
    expense_account_id: int | None
    is_active: bool
    archived_at: datetime | None
    created_at: datetime
    updated_at: datetime

File: Frontend/src/types/product.ts
export type ProductType = 'GOODS' | 'SERVICE' | 'COMBO'

export interface ProductCategory {
  id: number
  name: string
  is_active: boolean
}

export interface Product {
  id: number
  name: string
  product_type: ProductType
  category_id: number | null
  category: ProductCategory | null
  sales_price: string
  purchase_cost: string
  default_tax_rate: string
  income_account_id: number | null
  expense_account_id: number | null
  is_active: boolean
  archived_at: string | null
  created_at: string
  updated_at: string
}

export interface ProductInput {
  name: string
  product_type: ProductType
  category_id?: number | null
  sales_price?: string
  purchase_cost?: string
  default_tax_rate?: string
  income_account_id?: number | null
  expense_account_id?: number | null
}

export const PRODUCT_TYPE_LABELS: Record<ProductType, string> = {
  GOODS: 'Goods',
  SERVICE: 'Service',
  COMBO: 'Combo',
}
